autotest/TestList.f90

```diff
@@ -0,0 +1,254 @@
+module TestList
+  use KindModule, only: I4B
+  use testdrive, only: error_type, unittest_type, new_unittest, check
+  use ConstantsModule, only: LINELENGTH
+  use ListModule, only: ListType
+  implicit none
+  private
+  public :: collect_list
+
+  type :: IntNodeType
+    integer :: value
+  end type IntNodeType
+
+contains
+
+  subroutine collect_list(testsuite)
+    type(unittest_type), allocatable, intent(out) :: testsuite(:)
+    testsuite = [ &
+                new_unittest("add_count_get_item", &
+                             test_add_count_get_item), &
+                new_unittest("get_get_index_contains", &
+                             test_get_index_contains), &
+                new_unittest("get_next_previous_item_reset", &
+                             test_get_next_previous_item_reset), &
+                new_unittest("insert_after", &
+                             test_insert_after), &
+                new_unittest("remove_node", &
+                             test_remove_node) &
+                ]
+  end subroutine collect_list
+
+  subroutine test_add_count_get_item(error)
+    type(error_type), allocatable, intent(out) :: error
+    type(ListType), pointer :: list
+    type(IntNodeType), pointer :: n
+    class(*), pointer :: p
+
+    allocate (list)
+    allocate (n)
+
+    ! empty
+    call check(error, list%Count() == 0, "count should be 0")
+    if (allocated(error)) return
+
+    ! add one node
+    n%value = 1
+    p => n
+    call list%Add(p)
+
+    ! check count
+    call check(error, list%Count() == 1, "count should be 1")
+    if (allocated(error)) return
+
+    ! retrieve item
+    p => list%GetItem(1)
+    call check(error, associated(p, n))
+    select type (item => p)
```

Comment:
I'm not sure i've seen this syntax before.  Item is not declared, but yet it is assigned the correct type.